docs/source/contribution/cpu_features.rst

```diff
@@ -0,0 +1,214 @@
+                       ``float`` or ``double``.
+- ``method`` Computational methods of the algorithm. ``method1`` or ``method2`` in the case of ``Abc``.
+- ``cpu`` Version of the cpu-specific implementation of the algorithm, ``daal::CpuType``.
+
+Implementations for different methods are usually defined usind partial class templates specialization.
+
+\*_impl.i
+---------
+
+Those files contain the implementations of the computational functions defined in `*_kernel.h` files.
+Here is a variant of ``method1`` imlementation for ``Abc`` training algorithm that does not contain any
+instruction set specific code. The implementation is located in the file `abc_classification_train_method1_impl.i`:
+
+.. include:: ../includes/cpu_features/abc-classification-train-method1-impl.rst
+
+Although the implementation of the ``method1`` does not contain any instruction set specific code, it is
+expected that the developers leverage SIMD related macros available in oneDAL.
+For example, ``PRAGMA_IVDEP``, ``PRAGMA_VECTOR_ALWAYS``, ``PRAGMA_VECTOR_ALIGNED`` and others pragmas defined in
+`service_defines.h <https://github.com/oneapi-src/oneDAL/blob/main/cpp/daal/src/services/service_defines.h>`_.
+This will guide the compiler to generate more efficient code for the target architecture.
+
+Consider that the implementation of the ``method2`` for the same algorithm will be different and will contain
+AVX-512-specific code located in ``cpuSpecificCode`` function.
+Then the implementation of the ``method2`` in the file `abc_classification_train_method2_impl.i` will look like:
+
+.. include:: ../includes/cpu_features/abc-classification-train-method2-impl.rst
+
+CPU-specific code needs to be placed under compiler-specific and CPU-specific defines because it usually
+contains intrinsics that cannot be compiled on other architectures.
+
+\*_fpt_cpu.cpp
+--------------
+
+Those files contain the instantiations of the template classes defined in `*_kernel.h` files.
+The instatiation of the ``Abc`` training algorithm kernel for ``method1`` is located in the file
+`abc_classification_train_method1_batch_fpt_cpu.cpp`:
+
+.. include:: ../includes/cpu_features/abc-classification-train-method1-fpt-cpu.rst
+
+`_fpt_cpu.cpp` files are not compiled directly into object files. First, multiple copies of those files
+are made replacing the ``fpt`` and ``cpu`` parts of the file name as well as the corresponding ``DAAL_FPTYPE`` and
+``DAAL_CPU`` macros with the actual data type and CPU type values. Then the resulting files are compiled
+with appropriate CPU-specific optimization compiler options.
+
+The values for ``fpt`` file name part replacement are:
+- ``flt`` for ``float`` data type, and
+- ``dbl`` for ``double`` data type.
+
+The values for ``DAAL_FPTYPE`` macro replacement are ``float`` and ``double`` respectively.
+
+The values for ``cpu`` file name part replacement are:
+- ``nrh`` for Intel |reg| SSE2 architecture, which stands for Northwood,
```

Comment:
What's the difference between these and `DAAL_CPU`? Is there some case in which they wouldn't have a 1-to-1 match? When should these be preferred over `DAAL_CPU`?

Reply:
There is really no big difference. They are basically describe the same parameter of the resulting product.
`cpu` is a part of the file names - matters to build system, and `DAAL_CPU` is a macro which is a part of the code.
Of course, it is possible to have those two (`cpu` and `DAAL_CPU`) in alignment. But this chapter describes how it is done now.

I do not know exactly, but it might be that those two values were made different intentionally.
I can imagine two different CPUs that share the same instruction set, but have different througputs/latencies of the instructions; so the compiler might provide different optimization option for those CPUs.
In this case it would be beneficial to differentiate between `cpu` and `DAAL_CPU` variables.